apps/autoscheduler/AutoSchedule.cpp

```diff
@@ -0,0 +1,3296 @@
+/*
+  This file is the core of the autoscheduler. Most of the code here is
+  about navigating the search space and computing the
+  featurization. This also contains the top-level interface into the
+  autoscheduler.
+
+  The most interesting classes to look are:
+
+  LoopNest               Represents one node in our tree representation of loop nests.
+  State                  A state in the beam search. Holds a root loop nest.
+
+  Interesting functions below are:
+
+  generate_schedule            The top-level entrypoint, which computes and applies a schedule to a Halide pipeline
+  optimal_schedule             Runs the passes of the coarse-to-fine beam search
+  optimal_schedule_pass        Runs a single pass of beam search
+  LoopNest::compute_features   Recursively walks over a loop nest tree, computing our featurization using Halide's analysis tools.
+  LoopNest::apply              Actually applied a computed schedule to a Halide pipeline
+  State::generate_children     Generators successor states to a state in the beam search
+
+  Environment variables used (directly or indirectly):
+
+  HL_BEAM_SIZE
+  Beam size to use in the beam search. Defaults to 32. Use 1 to get a greedy search instead.
+
+  HL_CYOS
+  "Choose-your-own-schedule". If set to 1, lets you navigate the search tree by hand in the terminal. Whee! This is for debugging the autoscheduler.
+
+  HL_FEATURE_FILE -> output
+  Write out a training sample for the selected schedule into this file. Needs to be augmented with the runtime using augment_sample before it can be used to train.
+
+  HL_MACHINE_PARAMS
+  An architecture description string. Used by Halide master to configure the cost model. We only use the first term. Set it to the number of cores to target.
+
+  HL_PERMIT_FAILED_UNROLL
+  Set to 1 to tell Halide not to freak out if we try to unroll a loop that doesn't have a constant extent. Should generally not be necessary, but sometimes the autoscheduler's model for what will and will not turn into a constant during lowering is inaccurate, because Halide isn't perfect at constant-folding.
+
+  HL_SCHEDULE_FILE
+  Write out a human-and-machine readable block of scheduling source code for the selected schedule into this file.
+
+  HL_RANDOM_DROPOUT
```

Comment:
I think we need some way to pass arbitrary arguments to generator plugins (-p). Any thoughts? We could do something like let them sniff and mutate argc and argv directly, or we could let them register extra GeneratorParams somehow, though it's per-process, not per-generator, so that would be awkward. Or we could have a special flag that means "here's an argument to a plugin", but then if you have multiple you need to specify which...